### PR TITLE
gh-80145: Exclude '.' directory when adding directories to a tar archive

### PR DESCRIPTION
**Pull request by websurfer5**
As a result of the current tarfile behavior, shutil.make_archive (xxx, tar, root_dir) results in a './' entry in the archive, which is wrong.

* Issue: gh-80145

### CODE REVIEW

**Review by auvipy — APPROVED** [review]

File: Lib/tarfile.py [hunk 1]
```diff
@@ -1969,11 +1969,13 @@ def add(self, name, arcname=None, recursive=True, *, filter=None):
                 self.addfile(tarinfo, f)
 
         elif tarinfo.isdir():
-            self.addfile(tarinfo)
+            if tarinfo.name != '.':
```
Comment (taleinat) [inline Lib/tarfile.py]: I'm not deeply familiar enough with tar to know the answer to this, but should this be hard-coded to `'.'` or use `os.curdir`? (here and below as well)

**Review by taleinat — CHANGES_REQUESTED** [review]
I don't think the change to shutil warrants a mention in "What's New". The similar fix for zip archives did not add such a mention. The "What's New" line about the change to tarfile is fine IMO, though.

Also, since this changes the tarfile module (rather than shutil), this also should be tested in the tar module's tests.

File: Lib/tarfile.py (hunk 1 above)
Comment (websurfer5) [reply Lib/tarfile.py]: I have made the requested changes; please review again

File: Lib/tarfile.py [hunk 2]
```diff
@@ -1969,11 +1969,13 @@ def add(self, name, arcname=None, recursive=True, *, filter=None):
...
+                self.addfile(tarinfo)
             if recursive:
                 for f in sorted(os.listdir(name)):
-                    self.add(os.path.join(name, f), os.path.join(arcname, f),
-                            recursive, filter=filter)
+                    if f != '.':
```
Comment (taleinat) [inline Lib/tarfile.py]: Is this necessary? I don't believe `os.listdir` will ever include `'.'` in its results.

Comment (websurfer5) [reply Lib/tarfile.py]: I have made the requested changes; please review again

**Review by taleinat — APPROVED** [review]
Looks good to me.

This should also be reviewed by @gustaebel, the expert on the tarfile module.

**Comment by MadhavEsDios** [comment]
I would really like for this simple change to be merged. Thanks @websurfer5 
I am currently facing the same issue of the additional './' while trying to create a tar file using shutil.make_archive()

File: Lib/test/test_tarfile.py [hunk 3]
```diff
@@ -1493,8 +1493,9 @@ def test_cwd(self):
             tar = tarfile.open(tmpname, "r")
             try:
                 for t in tar:
-                    if t.name != ".":
-                        self.assertTrue(t.name.startswith("./"), t.name)
+                    # bpo-35964: "." is now excluded from tarfiles
```
Comment (terryjreedy) [inline Lib/test/test_tarfile.py]:
```suggestion
                    # gh-80145: "." is now excluded from tarfiles
```

**Comment by terryjreedy** [comment]
I changes the title to reflect what the patch does, as opposed the the symption it prevents.  This raises the question of whether it is *always* wrong for tarfile to add './', so that this is truly a bug fix, of only when tarfile is called from shutil?  In other words, is the bug possibly in the arguments passed by shutil?  Tal Einat's approval of this PR was conditioned on an answer from a 'tarfile expert'.  Lars is no longer reviewing.  @serhiy-storchaka Can you decide this?

Otherwise, if the tests still pass after merging main, I think this good to go.  The OP changed test_tarfile as Tal requested.  I removed the 3.11 backport label and updated a comment from 'bpo' to 'gh'.

**Comment by gvanrossum** [comment]
Well, it would leave out the permissions ("mode") and other metadata for the current directory, which might be used when extracting (possibly using a different tar implementation) into a specific directory that doesn't exist yet but which that utility might create, setting the mode from the tarchive? So it seems to me this might not be 100% backwards compatible.

**Comment by nineteendo** [comment]
Isn't that the same as for zip archives? #72674

**Comment by serhiy-storchaka** [comment]
It is special because we have no way to create a tar file without the `'.'` entry and the `./` prefix. With the `tar` utility we can at least call `tar *` instead of `tar .`.

**Comment by gvanrossum** [comment]
Surely we can do the equivalent of `*` using the `glob` module?

**Comment by serhiy-storchaka** [comment]
`os.listdir()` is simpler. We can do this inside of `shutil.make_archive()`, currently the user of `shutil.make_archive()` cannot affect this. This is what #118152 does -- changes the default behavior (to presumably more useful), while leaving the possibility to get the old result if the user really needs it.

**Pull request closed without merging**